histogram.go

```diff
@@ -0,0 +1,254 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tally
+
+import (
+	"fmt"
+	"math"
+	"sort"
+	"time"
+)
+
+// Defaults is a type that can represent a constant.
+type Defaults int
+
+// DefaultBuckets can be passed to specify to default buckets.
+const DefaultBuckets = Defaults(0)
+
+func (v Defaults) String() string {
+	return "[defaults]"
+}
+
+// Len implements Buckets.
+func (v Defaults) Len() int { return 0 }
+
+// AsValues implements Buckets.
+func (v Defaults) AsValues() []float64 { return nil }
+
+// AsDurations implements Buckets.
+func (v Defaults) AsDurations() []time.Duration { return nil }
+
+// Values is a set of float64 values that implements Buckets.
+type Values []float64
+
+// Implements sort.Interface
+func (v Values) Len() int {
+	return len(v)
+}
+
+// Implements sort.Interface
+func (v Values) Swap(i, j int) {
+	v[i], v[j] = v[j], v[i]
+}
+
+// Implements sort.Interface
+func (v Values) Less(i, j int) bool {
+	return v[i] < v[j]
+}
+
+func (v Values) String() string {
+	values := make([]string, len(v))
+	for i := range values {
+		values[i] = fmt.Sprintf("%f", v[i])
+	}
+	return fmt.Sprintf("%v", values)
+}
+
+// AsValues implements Buckets.
+func (v Values) AsValues() []float64 {
+	sort.Sort(v) // Always sort first
+	return []float64(v)
+}
+
+// AsDurations implements Buckets.
+func (v Values) AsDurations() []time.Duration {
+	sort.Sort(v) // Always sort first
+	values := make([]time.Duration, len(v))
+	for i := range values {
+		values[i] = time.Duration(v[i] * float64(time.Second))
+	}
+	return values
+}
+
+// Durations is a set of float64 values that implements Buckets.
+type Durations []time.Duration
+
+// Implements sort.Interface
+func (v Durations) Len() int {
+	return len(v)
+}
+
+// Implements sort.Interface
+func (v Durations) Swap(i, j int) {
+	v[i], v[j] = v[j], v[i]
+}
+
+// Implements sort.Interface
+func (v Durations) Less(i, j int) bool {
+	return v[i] < v[j]
+}
+
+func (v Durations) String() string {
+	sort.Sort(v) // Always sort first
+	values := make([]string, len(v))
+	for i := range values {
+		values[i] = v[i].String()
+	}
+	return fmt.Sprintf("%v", values)
+}
+
+// AsValues implements Buckets.
+func (v Durations) AsValues() []float64 {
+	sort.Sort(v) // Always sort first
+	values := make([]float64, len(v))
+	for i := range values {
+		values[i] = float64(v[i]) / float64(time.Second)
+	}
+	return values
+}
+
+// AsDurations implements Buckets.
+func (v Durations) AsDurations() []time.Duration {
+	sort.Sort(v) // Always sort first
+	return []time.Duration(v)
+}
+
+// BucketPairs creates a set of bucket pairs from a set
+// of buckets describing the lower and upper bounds for
+// each derived bucket.
+func BucketPairs(buckets Buckets) []BucketPair {
+	var (
+		asValueBuckets    = buckets.AsValues()
+		asDurationBuckets = buckets.AsDurations()
+	)
+	var pairs []BucketPair
```

Comment:
It's not necessarily the same length, in fact a lot of the time there's at least one added pair to add the range `[-infinity, first_bucket)` and `(last_bucket, infinity]`.